File: 1368-minimum-cost-to-make-at-least-one-valid-path-in-a-grid/1368-minimum-cost-to-make-at-least-one-valid-path-in-a-grid.cpp
class Solution {
public:
    vector<vector<int>> dirs = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    int m, n;
    vector<vector<bool>> vis;

    int dfs(int i, int j, vector<vector<int>> &grid, int cost) {
        if (i == m - 1 && j == n - 1) {
            return cost;
        }
        vis[i][j] = true;
        int minCost = INT_MAX;
        for (int dir = 0; dir < 4; dir++) {
            int x = i + dirs[dir][0];
            int y = j + dirs[dir][1];
            if (x >= 0 && y >= 0 && x < m && y < n && !vis[x][y]) {
                int nextCost = cost + ((grid[i][j] - 1 != dir) ? 1 : 0);
                minCost = min(minCost, dfs(x, y, grid, nextCost));
            }
        }
        vis[i][j] = false;
        return minCost;
    }

    int minCost(vector<vector<int>>& grid) {
        m = grid.size();
        n = grid[0].size();
        vis.resize(m, vector<bool>(n, false));
        return dfs(0, 0, grid, 0);
    }
};

File: 1368-minimum-cost-to-make-at-least-one-valid-path-in-a-grid/1368-minimum-cost-to-make-at-least-one-valid-path-in-a-grid.py
class Solution:
    def minCost(self, grid: List[List[int]]) -> int:
        dirs = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        m, n = len(grid), len(grid[0])
        pq = []
        heappush(pq, (0, 0, 0))
        res = [[float('inf')] * n for _ in range(m)]
        res[0][0] = 0
        
        while pq:
            currCost, i, j = heappop(pq)
            if res[i][j] < currCost:
                continue
            for dir, (dx, dy) in enumerate(dirs):
                x, y = i + dx, j + dy
                if 0 <= x < m and 0 <= y < n:
                    gridDir = grid[i][j]
                    dirCost = 1 if gridDir - 1 != dir else 0
                    newCost = currCost + dirCost
                    if newCost < res[x][y]:
                        res[x][y] = newCost
                        heappush(pq, (newCost, x, y))
        
        return res[m - 1][n - 1]